crates/subspace-archiving/src/piece_reconstructor.rs

```diff
@@ -0,0 +1,225 @@
+extern crate alloc;
+
+use crate::utils;
+use alloc::vec::Vec;
+use reed_solomon_erasure::galois_16::ReedSolomon;
+use subspace_core_primitives::crypto::blake2b_256_254_hash;
+use subspace_core_primitives::crypto::kzg::{Kzg, Polynomial};
+use subspace_core_primitives::{
+    FlatPieces, Piece, BLAKE2B_256_HASH_SIZE, PIECES_IN_SEGMENT, PIECE_SIZE,
+};
+
+/// Reconstructor-related instantiation error.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+#[cfg_attr(feature = "thiserror", derive(thiserror::Error))]
+pub enum ReconstructorInstantiationError {
+    /// Segment size is not bigger than record size
+    #[cfg_attr(
+        feature = "thiserror",
+        error("Segment size is not bigger than record size")
+    )]
+    SegmentSizeTooSmall,
+    /// Segment size is not a multiple of record size
+    #[cfg_attr(
+        feature = "thiserror",
+        error("Segment size is not a multiple of record size")
+    )]
+    SegmentSizesNotMultipleOfRecordSize,
+}
+
+/// Reconstructor-related instantiation error
+#[derive(Debug, Clone, PartialEq)]
+#[cfg_attr(feature = "thiserror", derive(thiserror::Error))]
+pub enum ReconstructorError {
+    /// Segment size is not bigger than record size
+    #[cfg_attr(
+        feature = "thiserror",
+        error("Error during data shards reconstruction: {0}")
+    )]
+    DataShardsReconstruction(reed_solomon_erasure::Error),
+
+    /// Incorrect piece position provided.
+    #[cfg_attr(feature = "thiserror", error("Incorrect piece position provided."))]
+    IncorrectPiecePosition,
+}
+
+/// Reconstructor helps to retrieve blocks from archived pieces.
+#[derive(Debug, Clone)]
+pub struct PiecesReconstructor {
+    /// Configuration parameter defining the size of one record (data in one piece excluding witness
+    /// size)
+    record_size: u32,
+    /// Configuration parameter defining the size of one recorded history segment
+    segment_size: u32,
+    /// Erasure coding data structure
+    reed_solomon: ReedSolomon,
+    /// KZG instance
+    kzg: Kzg,
+}
+
+impl PiecesReconstructor {
+    fn shards_count(segment_size: u32, record_size: u32) -> u32 {
+        segment_size / record_size
+    }
+
+    fn data_shards(&self) -> u32 {
+        Self::shards_count(self.segment_size, self.record_size)
+    }
+
+    fn parity_shards(&self) -> u32 {
+        self.data_shards()
```

Comment:
They happen to be the same, but they don't have to be. This is why archiver had them explicitly separated.